torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,172 @@
+#include "nccl.h"
+#include "torch/csrc/THP.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/cuda/THCP.h"
+
+#include <nccl.h>
+#include <sstream>
+#include <unordered_map>
+
+static inline void CHECK(ncclResult_t status)
+{
+  if (status != ncclSuccess) {
+    std::stringstream err;
+    err << "NCCL Error " << status << ": " << ncclGetErrorString(status);
+    throw std::runtime_error(err.str());
+  }
+}
+
+struct NcclCommList {
+  std::unique_ptr<ncclComm_t[]> comms;
+  int ndevices;
+  NcclCommList(const std::vector<int>& devices)
+    : comms(new ncclComm_t[devices.size()]), ndevices(devices.size()) {
+    CHECK(ncclCommInitAll(comms.get(), devices.size(), devices.data()));
+  }
+  NcclCommList(NcclCommList&& foo) = default;
+  ~NcclCommList() {
+    if (comms) {
+      for (int i = 0; i < ndevices; i++) {
+	ncclCommDestroy(comms[i]);
```

Comment:
Considering that you have to call this per-elemt I doubt this calls `free` on the memory you allocated for the handles. You should still free it

Reply:
I think this is fine. The comms array is in a `unique_ptr`.